lib/puppet/provider/zfs_share/solaris.rb

```diff
@@ -0,0 +1,223 @@
+      allow_ip_read  = @resource[:allow_ip_read]
+      allow_ip_write = @resource[:allow_ip_write]
+      anon_user_id   = @resource[:anon_user_id]
+      nosub          = @resource[:nosub]
+      types          = ['none', 'sys', 'default']
+      if security.kind_of?(String)
+        security = [security]
+      end
+      types.reject!{ |x| security.include? x}
+      types.each do |t|
+        if share[/sec=#{t}/]
+          IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 3].join(',')},sec=#{t}", name])
+        end
+      end
+      sec = share[/sec=\w*(?=($|,))/]
+      if not allow_ip_read
+        if share =~ /,ro=[@|\*]/
+          IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 3].join(',')},#{sec},ro", name])
+        end
+      end
+      if not allow_ip_write
+        if share =~ /,rw=[@|\*]/
+          IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 3].join(',')},#{sec},rw", name])
+        end
+      end
+      if not nosub
+        if share =~ /nosub=true/
+          IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 3].join(',')},nosub", name])
+        end
+      end
+      if not anon_user_id
+        if share =~ /anon=\d+/
+          IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 3].join(',')},anon", name])
+        end
+      end
+      if protocol
+        if not protocol.include?('smb')
+          if share =~ /prot=smb/
+            IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 2].join(',')}prot=smb", name])
+          end
+        end
+      end
+    end
+  end
+
+  def purge_share
+    # "When the share to be set and the share on disk differ from one another first remove the share currently configured in its entirety.
+    # This allows changing the share title without manual intervention after initially setting it. The caveat of this parameter is that
+    # there's a potential for the NFS server connection on the remote end to be lost during the brief period that access is removed."
+    get_share = ['zfs', 'get', 'share', share_path()]
+    share = return_cmd(get_share)
+    if share.length > 0
+      share_split = share[-1].split()[2]
+      IO.popen(['zfs', 'set', '-c', "share=#{share_split.split(',')[0, 2].join(',')}", share_path()])
+    end
+  end
+
+  def exists?
+    # "Check is the share exists."
+    begin
+      name = share_path()
+      if File.directory?("/#{name}")
+        get_share_nfs = ['zfs', 'get', 'share.nfs', name]
+        get_share = ['zfs', 'get', 'share', name]
+        output = return_cmd(get_share_nfs)
+        share = return_cmd(get_share)
+        if output.length > 0
+          if output[-1].split()[2] == 'on'
+            if share.length > 0
+              check_for_stale_data()
+              share_ = share_cmd()
+              if share[-1].split()[2] == share_
+                true
+              end
+            end
+          end
+        end
+      end
+    end
+    rescue Puppet::ExecutionFailure
+      false
+  end
+
+  def create
```

Comment:
Both of create and destroy are implementing the same pattern as the define type is, setting the share property.  We've already identified this is the incorrect way to configure sharing in Solaris 11.  You'll need to iterate over various type parameters and properties and do `zfs set` for all zfs properties, e.g. `share.nfs.sec.none.rw`.